### DISCUSSION TITLE
Migrating from tfussell's repo to xlnt-community

### DISCUSSION THREAD

**Discussion by m7913d** [xlnt-community/xlnt#4] [General]
To make this repo a success, it seems important to me that others also start using this repo for PR and reporting of issues.

Should we actively monitoring tfussell's repo for a while and notify them of the currently unmaintained state of tfussell's repo and of the existence of this community effort to continue his work?

@doomlaur What do you think?

**Comment by m7913d** [comment]
My suggestion is to post the following message on the now unmaintained XLNT repo of tfussell for new and recent issues and PRs:

> As discussed in issue #748, this repo has been unmaintained for several years.
> In order to continue this great project started by tfussell, we have created a [new repo](https://github.com/xlnt-community/xlnt) to  support further development by the community.
> Feel free to participate in this community effort by submitting issues and PR to this new community-driven repo.

@doomlaur What do you think?

**Comment by doomlaur** [comment]
Sorry for the delay :sweat_smile: I think it's a good idea, yes :smile: This is probably the best we can do at the moment, as we cannot change the `README.md` or pin an issue in the original repository.

By the way: at some point we should also think about having a documentation page, like the [original XLNT documentation](https://tfussell.gitbooks.io/xlnt/content/). For now it's not that important yet, as both repositories have the same features, but in time that will change.

**Comment by m7913d** [comment]
Also the README file should receive an update in the near future:
 - describe relation between this repo and tfussell's one
 - update the build/coverage status at the top of the README file to link to the actual status of our repo.

**Comment by m7913d** [comment]
I also think about hosting the API reference generated by Doxygen (using github pages?). The gitbooks seems fine for handling some important/getting started topics, but I don't like the formatting of the (incomplete) API reference in gitbooks.

**Comment by doomlaur** [comment]
Sounds very good! :+1:

**Comment by m7913d** [comment]
I think most of the migration process is realized. Two things we may still consider:
 1. unforking to become a stand-alone xlnt repo. In the end, I think this is the way to go, but maybe not yet. Disadvantage of being a fork is that PRs are automatically created against tfussells repo and not ours. Unforking may also make it clear that our repo exists independently of tfussells one. However, unforking has also some (short term) disadvantages: merging from forks of tfussells repo may become less trivial and we will not be listed anymore as an active fork of tfussells repo (which may make it harder to find our repo for existing xlnt users).
 2. Releasing a new version / using our repo when building using vcpkg (Note that I don't have any experience with vcpkg, not even as a user)

**Reply by doomlaur** [reply]
> 1. unforking to become a stand-alone xlnt repo. In the end, I think this is the way to go, but maybe not yet. Disadvantage of being a fork is that PRs are automatically created against tfussells repo and not ours. Unforking may also make it clear that our repo exists independently of tfussells one.

Talking with some work colleagues about XLNT, I found out that it was hard to find our XLNT fork. This is probably because GitHub wants the main repository at the top of the search results of the different search machines, instead of having many different small forks. Makes sense in general, but it's a disadvantage for us at the moment.

In other words, unless other people read your answers to the issues and PRs in tfussell's XLNT repository, it's almost impossible to find our repository. This could also explain the slow increase of stars we talked about in https://github.com/xlnt-community/xlnt/discussions/32#discussioncomment-11959576.

> However, unforking has also some (short term) disadvantages: merging from forks of tfussells repo may become less trivial

To my knowledge, we are the only fork that has any significant activity at all. All other forks only contain very small fixes needed for special cases, but otherwise no further issues or PRs. Of course, this could theoretically still change in the future.

> and we will not be listed anymore as an active fork of tfussells repo (which may make it harder to find our repo for existing xlnt users).

Agreed that this is a disadvantage - however, you also made it clear multiple times in tfussell's repo that the library is unmaintained and our fork is the currently active one. I think this should (hopefully) cover the people who would find our fork by looking on GitHub.

Otherwise, I think we are at a real disadvantage being so hard to find in the search machines at the moment. Unforking would probably make our repository more popular (and much easier to find).

**Reply by m7913d** [reply]
Ok for me to unfork.

**Reply by doomlaur** [reply]
Unfortunately, while this is possible, the only way to do it without losing PRs and issues is by pressing `Leave the fork network` in the Settings (see [Detaching a fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/detaching-a-fork)), which itself is only possible if the repository has no forks. However, at this moment we already have 5 forks. Unless each of the owners either delete their repositories, or unfork themselves, it's impossible for us to unfork.

Unless we want to ask everyone to delete their repositories, or unless we delete our own (very bad idea), we are stuck like this.

From what I found on the internet, the option to leave the fork network was not always available, and back in the day people used to contact GitHub directly, or use its virtual assistant to do the same. This worked nicely, although no one wrote if they already had forks - so I have no idea if they would do it for us.

**Reply by m7913d** [reply]
I have submitted a request to GitHub support to detach our fork.

**Reply by doomlaur** [reply]
Not sure when it happened, but it seems we are detached now! :tada: :smile:

**Comment by MyRoGeertGit** [comment]
Hello,

On https://github.com/xlnt-community/xlnt/pulls the 'New Pull Request' button refers to tfussell's version.

Thank you for your efforts.
Geert.

**Comment by doomlaur** [comment]
@m7913d You are right and have summarized the advantages and disadvantages very well. I would propose the following:

1. On my TODO list I still have to fix the issue from #685, but we are now also discussing it in #18, so let's see how that evolves.
2. We should also think whether any of the issues from #644 are still relevant for us - they were being considered by @tfussell as requirements for releasing XLNT 1.6.
3. Think about whether it makes sense to continue under the simple name "XLNT" or whether we should add some prefix, suffix, or and modification of the library's name. If the original XLNT would still be developed, I would definitely propose to change the name (at least slightly, using a prefix or suffix), but in our current situation, we might decide against it. I'm just not sure whether we might have any problems in the long run (e.g. with vcpkg).
4. Release a new version of the library. This is very important for many people, as many people prefer to use stable versions. XLNT 1.5 still contains many issues that have been solved, so releasing version 1.6 would be very important.
5. Update vcpkg to the newest version - especially because I saw that many people used vcpkg and did not even notice that compiling XLNT from the master branch actually was more stable than version 1.5. However, I also don't have any experience with it, neither as a maintainer, nor as a user. I guess we'll have to do some research :wink:

**Comment by m7913d** [comment]
@doomlaur I think it may be a good time to create our first release:
- If you want some issues being fixed before the release. I think we should create a milestone for it. However, I would keep the number of included issues as low as possible.
- Concerning the name: I would like to keep the name as is. This would ensures vcpkg users just gets the new release. However, if vcpk doesn't accept this, I would use a name like xlnt-ce or xlnt community edition.

**Comment by doomlaur** [comment]
@m7913d I would like to do the following before the release:

1. ~~Fix the unit tests for MSVC due to January 1st 1970 (as explained in #22).~~ Done in https://github.com/xlnt-community/xlnt/pull/27.
2. ~~Improve the date/time functions and add them to some utils function (as explained in #22).~~ Done in https://github.com/xlnt-community/xlnt/pull/27.
3. ~~Replace `std::codecvt` due to deprecation in C++17 and removal in C++26.~~ Done in https://github.com/xlnt-community/xlnt/pull/29.
4. ~~Update copyright information, as explained in #24.~~ Done in https://github.com/xlnt-community/xlnt/pull/30.
5. ~~Provide an improved parser for floating-point numbers that no longer requires using `std::localeconv`, or (for the unit tests)) `std::setlocale`, as explained in the discussions from #2~~ Done in https://github.com/xlnt-community/xlnt/pull/38.

I'll try to get everything done today in the evening (since I'm away from home the whole day). Sorry for the delay in finishing the things listed above - I have some important things going on (see my website for more infos, if you are interested).

EDIT: I updated the progress - I'll finish the rest tomorrow morning :wink:

**Comment by doomlaur** [comment]
A small update from my side, in addition to the updated comment above: number 5 from my list (improved parser, avoiding `std::localeconv` and `std::setlocale`) is mostly done (and compiles again), however, I would like to test it a bit more to make sure that it doesn't cause any issues. Such changes to serialization and deserialization can always cause issues when not being careful, and I would definitely like to avoid that. However, this change is NOT absolutely necessary for the next minor version, so if you would prefer to release it without it, I wouldn't mind either.

There's one more thing I noticed: while you updated `utfcpp` and I updated `miniz`, and while `libstudxml` is almost entirely up to date (apart from very few insignificant commits), the one library that is rather old is `pybind11`. I did not try to update it yet, but noticed that there were quite a few security reports related to it: https://security.snyk.io/vuln/?search=pybind

I'm not sure if the security reports are relevant to us in any way (I did not have time to check it out yet), but we should either update the library (if easily possible), or plan to do so very soon. Actually finding out whether the security reports are relevant is probably more work than just updating the library.

**Comment by m7913d** [comment]
If it is error prone, it may be better to keep it out of the release for now. This makes the release somewhat more stable.

Ok for updating the pybind library. However, I don't use python and I think no unit tests are available for python. So, we may need to perform some basic testing to ensure at least something works with the updated library. I don't know how important this binding is, as I would expect python having its own xlsx reader/writer.

**Comment by doomlaur** [comment]
I sent a PR for updating pybind11 in #31.

Theoretically, all changes to serialization (formatting) and deserialization (parsing) can be error prone if not tested very carefully. While I would like to add the changes, I can argue that:

1. The changes are not critical at the moment, as you already fixed some important issues in #2.
2. I won't have much time over the next couple of days, so especially because of this, I would prefer to not send a rushed PR just to get it done, potentially at the risk of breaking something.
3. While you could wait for me to get it done and properly tested, I don't think the parser changes are critical for the next minor version. We can wait for me to get it done and tested, but don't have to.

One more thing before releasing a new version: I noticed that there are some last links that we did not update yet with the information of our fork:

- `package.json` is used, as far as I see, to release the [XLNT npm package](https://www.npmjs.com/package/xlnt) which allows others to use XLNT in the browser in WebAssembly. This file contains information that is out of date, and the two dependencies it has are also out of date.
- `python/setup.py` also contains information that is out of date, which is used to update the [`xlntpyarrow` package](https://anaconda.org/conda-forge/xlntpyarrow).

**Comment by yang8570** [comment]
잘사용하고 있습니다.

 format_impl *find_or_create(format_impl pattern)
   //     if (id != pattern.id)
   //     {
   //         iter = format_impls.begin();
   //         std::advance(iter, static_cast<std::list<format_impl>::difference_type>(pattern.id));
   //         iter->references -= iter->references > 0 ? 1 : 0;
			//garbage_collection_enabled = false;
			//garbage_collect();
   //     }


    format_impl *find_or_create_with(format_impl *pattern, const std::string &style_name)
    format_impl *find_or_create_with(format_impl *pattern, const alignment &new_alignment, optional<bool> applied)
    format_impl *find_or_create_with(format_impl *pattern, const border &new_border, optional<bool> applied)
    format_impl *find_or_create_with(format_impl *pattern, const border &new_border, optional<bool> applied)
        //if (pattern->references == 0)
        //{
        //    *pattern = new_format;
        //}


xlsx화일을 분석하여 보면은 동일한 항목이 같이 존재합니다.즉
    // Format XFs 항목은 중복되지 않는 항목이 아니고 적용한 히스토리를 계속 열거 하는 방식인것 같아요..
아래 처럼 수정했을때 원하는 목적을 쟁취했습니다...
감사합니다.





// Copyright (c) 2014-2021 Thomas Fussell
// Copyright (c) 2010-2015 openpyxl
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE
//
// @license: http://www.opensource.org/licenses/mit-license.php
// @author: see AUTHORS file
#pragma once

#include <list>
#include <string>
#include <vector>

#include <detail/implementations/conditional_format_impl.hpp>
#include <detail/implementations/format_impl.hpp>
#include <detail/implementations/style_impl.hpp>
#include <xlnt/cell/cell.hpp>
#include <xlnt/styles/conditional_format.hpp>
#include <xlnt/styles/format.hpp>
#include <xlnt/styles/style.hpp>
#include <xlnt/utils/exceptions.hpp>
#include <xlnt/workbook/workbook.hpp>
#include <xlnt/workbook/worksheet_iterator.hpp>
#include <xlnt/worksheet/worksheet.hpp>

namespace xlnt {
namespace detail {

struct stylesheet
{
    class format create_format(bool default_format)
    {
		xlnt::detail::format_impl newfm = format_impl();
		newfm.alignment_id.clear();
		newfm.alignment_applied.clear();
		newfm.border_applied.clear();
		newfm.border_id.clear();
		newfm.fill_applied.clear();
		newfm.fill_id.clear();
		newfm.font_applied.clear();
		newfm.font_id.clear();
		newfm.number_format_applied.clear();
		newfm.number_format_id.clear();
		newfm.protection_applied.clear();
		newfm.protection_id.clear();
		newfm.style.clear();

		format_impls.push_back(newfm);
		auto &impl = format_impls.back();

		impl.parent = this;
		impl.id = format_impls.size() - 1;

        impl.references = default_format ? 1 : 0;
        
        return xlnt::format(&impl);
    }

    class xlnt::format format(std::size_t index)
    {
        auto iter = format_impls.begin();
        std::advance(iter, static_cast<std::list<format_impl>::difference_type>(index));

        return xlnt::format(&*iter);
    }

    class style create_style(const std::string &name)
    {
        auto &impl = style_impls.emplace(name, style_impl()).first->second;

		impl.parent = this;
        impl.name = name;

        impl.border_id = 0;
        impl.fill_id = 0;
        impl.font_id = 0;
        impl.number_format_id = 0;

        style_names.push_back(name);
        return xlnt::style(&impl);
    }

	class style create_builtin_style(const std::size_t builtin_id)
	{
		// From Annex G.2
		static const auto names = std::unordered_map<std::size_t , std::string>
		{
			{ 0, "Normal" },
			{ 1, "RowLevel_1" },
			{ 2, "ColLevel_1" },
			{ 3, "Comma" },
			{ 4, "Currency" },
			{ 5, "Percent" },
			{ 6, "Comma [0]" },
			{ 7, "Currency [0]" },
			{ 8, "Hyperlink" },
			{ 9, "Followed Hyperlink" },
			{ 10, "Note" },
			{ 11, "Warning Text" },
			{ 15, "Title" },
			{ 16, "Heading 1" },
			{ 17, "Heading 2" },
			{ 18, "Heading 3" },
			{ 19, "Heading 4" },
			{ 20, "Input" },
			{ 21, "Output"},
			{ 22, "Calculation"},
			{ 22, "Calculation" },
			{ 23, "Check Cell" },
			{ 24, "Linked Cell" },
			{ 25, "Total" },
			{ 26, "Good" },
			{ 27, "Bad" },
			{ 28, "Neutral" },
			{ 29, "Accent1" },
			{ 30, "20% - Accent1" },
			{ 31, "40% - Accent1" },
			{ 32, "60% - Accent1" },
			{ 33, "Accent2" },
			{ 34, "20% - Accent2" },
			{ 35, "40% - Accent2" },
			{ 36, "60% - Accent2" },
			{ 37, "Accent3" },
			{ 38, "20% - Accent3" },
			{ 39, "40% - Accent3" },
			{ 40, "60% - Accent3" },
			{ 41, "Accent4" },
			{ 42, "20% - Accent4" },
			{ 43, "40% - Accent4" },
			{ 44, "60% - Accent4" },
			{ 45, "Accent5" },
			{ 46, "20% - Accent5" },
			{ 47, "40% - Accent5" },
			{ 48, "60% - Accent5" },
			{ 49, "Accent6" },
			{ 50, "20% - Accent6" },
			{ 51, "40% - Accent6" },
			{ 52, "60% - Accent6" },
			{ 53, "Explanatory Text" }
		};

		auto new_style = create_style(names.at(builtin_id));
		new_style.d_->builtin_id = builtin_id;

		return new_style;
	}

    class style style(const std::string &name)
	{
        if (!has_style(name)) throw key_not_found();
        return xlnt::style(&style_impls[name]);
	}

	bool has_style(const std::string &name)
	{
		return style_impls.count(name) > 0;
	}

	std::size_t next_custom_number_format_id() const
	{
		std::size_t id = 164;

		for (const auto &nf : number_formats)
		{
			if (nf.id() >= id)
			{
				id = nf.id() + 1;
			}
		}

		return id;
	}
    
    template<typename T, typename C>
    std::size_t find_or_add(C &container, const T &item)
    {
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wsign-conversion"
        auto iter = std::find(container.begin(), container.end(), item);
        if (iter != container.end())
        {
            return std::size_t(iter - container.begin());
        }
        iter = container.emplace(container.end(), item);
        return std::size_t(iter - container.begin());
#pragma GCC diagnostic pop
    }
    
    template<typename T>
    std::unordered_map<std::size_t, std::size_t> garbage_collect(
        const std::unordered_map<std::size_t, std::size_t> &reference_counts,
        std::vector<T> &container)
    {
        std::unordered_map<std::size_t, std::size_t> id_map;
        std::size_t unreferenced = 0;
        const auto original_size = container.size();

        for (std::size_t i = 0; i < original_size; ++i)
        {
            id_map[i] = i - unreferenced;

            if (reference_counts.count(i) == 0 || reference_counts.at(i) == 0)
            {
                container.erase(container.begin() + static_cast<typename std::vector<T>::difference_type>(i - unreferenced));
                unreferenced++;
            }
        }

        return id_map;
    }
    
    void garbage_collect()
    {
        if (!garbage_collection_enabled) return;
        
        auto format_iter = format_impls.begin();
        while (format_iter != format_impls.end())
        {
            auto &impl = *format_iter;

            if (impl.references != 0)
            {
                ++format_iter;
            }
            else
            {
                format_iter = format_impls.erase(format_iter);
            }
        }
        
        std::size_t new_id = 0;

        std::unordered_map<std::size_t, std::size_t> alignment_reference_counts;
        std::unordered_map<std::size_t, std::size_t> border_reference_counts;
        std::unordered_map<std::size_t, std::size_t> fill_reference_counts;
        std::unordered_map<std::size_t, std::size_t> font_reference_counts;
        std::unordered_map<std::size_t, std::size_t> number_format_reference_counts;
        std::unordered_map<std::size_t, std::size_t> protection_reference_counts;
        
        fill_reference_counts[0]++;
        fill_reference_counts[1]++;
        
        for (auto &impl : format_impls)
        {
            impl.id = new_id++;
            
            if (impl.alignment_id.is_set())
            {
                alignment_reference_counts[impl.alignment_id.get()]++;
            }
            
            if (impl.border_id.is_set())
            {
                border_reference_counts[impl.border_id.get()]++;
            }
            
            if (impl.fill_id.is_set())
            {
                fill_reference_counts[impl.fill_id.get()]++;
            }
            
            if (impl.font_id.is_set())
            {
                font_reference_counts[impl.font_id.get()]++;
            }
            
            if (impl.number_format_id.is_set())
            {
                number_format_reference_counts[impl.number_format_id.get()]++;
            }
            
            if (impl.protection_id.is_set())
            {
                protection_reference_counts[impl.protection_id.get()]++;
            }
        }
        
        for (auto &name_impl_pair : style_impls)
        {
            auto &impl = name_impl_pair.second;
            
            if (impl.alignment_id.is_set())
            {
                alignment_reference_counts[impl.alignment_id.get()]++;
            }
            
            if (impl.border_id.is_set())
            {
                border_reference_counts[impl.border_id.get()]++;
            }
            
            if (impl.fill_id.is_set())
            {
                fill_reference_counts[impl.fill_id.get()]++;
            }
            
            if (impl.font_id.is_set())
            {
                font_reference_counts[impl.font_id.get()]++;
            }
            
            if (impl.number_format_id.is_set())
            {
                number_format_reference_counts[impl.number_format_id.get()]++;
            }
            
            if (impl.protection_id.is_set())
            {
                protection_reference_counts[impl.protection_id.get()]++;
            }
        }
        
        auto alignment_id_map = garbage_collect(alignment_reference_counts, alignments);
        auto border_id_map = garbage_collect(border_reference_counts, borders);
        auto fill_id_map = garbage_collect(fill_reference_counts, fills);
        auto font_id_map = garbage_collect(font_reference_counts, fonts);
        auto protection_id_map = garbage_collect(protection_reference_counts, protections);

        for (auto &impl : format_impls)
        {
            if (impl.alignment_id.is_set())
            {
                impl.alignment_id = alignment_id_map[impl.alignment_id.get()];
            }
            
            if (impl.border_id.is_set())
            {
                impl.border_id = border_id_map[impl.border_id.get()];
            }
            
            if (impl.fill_id.is_set())
            {
                impl.fill_id = fill_id_map[impl.fill_id.get()];
            }
            
            if (impl.font_id.is_set())
            {
                impl.font_id = font_id_map[impl.font_id.get()];
            }
            
            if (impl.protection_id.is_set())
            {
                impl.protection_id = protection_id_map[impl.protection_id.get()];
            }
        }

        for (auto &name_impl : style_impls)
        {
            auto &impl = name_impl.second;

            if (impl.alignment_id.is_set())
            {
                impl.alignment_id = alignment_id_map[impl.alignment_id.get()];
            }
            
            if (impl.border_id.is_set())
            {
                impl.border_id = border_id_map[impl.border_id.get()];
            }
            
            if (impl.fill_id.is_set())
            {
                impl.fill_id = fill_id_map[impl.fill_id.get()];
            }
            
            if (impl.font_id.is_set())
            {
                impl.font_id = font_id_map[impl.font_id.get()];
            }
            
            if (impl.protection_id.is_set())
            {
                impl.protection_id = protection_id_map[impl.protection_id.get()];
            }
        }
    }

    format_impl *find_or_create(format_impl pattern)
    {
        pattern.references = 0;
        std::size_t id = 0;
        auto iter = format_impls.begin();
        while (iter != format_impls.end() /*&& !(*iter == pattern)*/)
        {
            ++id;
            ++iter;
        }
        if (iter == format_impls.end())
        {
            iter = format_impls.emplace(format_impls.end(), pattern);
        }
        auto &result = *iter;

        result.parent = this;
        result.id = id;
        result.references++;
        
   //     if (id != pattern.id)
   //     {
   //         iter = format_impls.begin();
   //         std::advance(iter, static_cast<std::list<format_impl>::difference_type>(pattern.id));
   //         iter->references -= iter->references > 0 ? 1 : 0;
			//garbage_collection_enabled = false;
			//garbage_collect();
   //     }

        return &result;
    }

    format_impl *find_or_create_with(format_impl *pattern, const std::string &style_name)
    {
        format_impl new_format = *pattern;
        new_format.style = style_name;
        //if (pattern->references == 0)
        //{
        //    *pattern = new_format;
        //}
        return find_or_create(new_format);
    }

    format_impl *find_or_create_with(format_impl *pattern, const alignment &new_alignment, optional<bool> applied)
    {
        format_impl new_format = *pattern;
        new_format.alignment_id = find_or_add(alignments, new_alignment);
        new_format.alignment_applied = applied;
        //if (pattern->references == 0)
        //{
        //    *pattern = new_format;
        //}
        return find_or_create(new_format);
    }

    format_impl *find_or_create_with(format_impl *pattern, const border &new_border, optional<bool> applied)
    {
        format_impl new_format = *pattern;
        new_format.border_id = find_or_add(borders, new_border);
        new_format.border_applied = applied;
        //if (pattern->references == 0)
        //{
        //    *pattern = new_format;
        //}
        return find_or_create(new_format);
    }
    
    format_impl *find_or_create_with(format_impl *pattern, const fill &new_fill, optional<bool> applied)
    {
        format_impl new_format = *pattern;
        new_format.fill_id = find_or_add(fills, new_fill);
        new_format.fill_applied = applied;
        //if (pattern->references == 0)
        //{
        //    *pattern = new_format;
        //}
        return find_or_create(new_format);
    }
    
    format_impl *find_or_create_with(format_impl *pattern, const font &new_font, optional<bool> applied)
    {
        format_impl new_format = *pattern;
        new_format.font_id = find_or_add(fonts, new_font);
        new_format.font_applied = applied;
        //if (pattern->references == 0)
        //{
        //    *pattern = new_format;
        //}
        return find_or_create(new_format);
    }
    
    format_impl *find_or_create_with(format_impl *pattern, const number_format &new_number_format, optional<bool> applied)
    {
        format_impl new_format = *pattern;
        //if (new_number_format.id() >= 164)
        //{
        //    find_or_add(number_formats, new_number_format);
        //}
        new_format.number_format_id = new_number_format.id();
        new_format.number_format_applied = applied;
        //if (pattern->references == 0)
        //{
        //    *pattern = new_format;
        //}
        return find_or_create(new_format);
    }
    
    format_impl *find_or_create_with(format_impl *pattern, const protection &new_protection, optional<bool> applied)
    {
        format_impl new_format = *pattern;
        new_format.protection_id = find_or_add(protections, new_protection);
        new_format.protection_applied = applied;
        //if (pattern->references == 0)
        //{
        //    *pattern = new_format;
        //}
        return find_or_create(new_format);
    }

    std::size_t style_index(const std::string &name) const
    {
        return static_cast<std::size_t>(std::distance(style_names.begin(),
	    std::find(style_names.begin(), style_names.end(), name)));
    }
    
    void clear()
    {
		conditional_format_impls.clear();
        format_impls.clear();
        
        style_impls.clear();
        style_names.clear();
        
        alignments.clear();
        borders.clear();
        fills.clear();
        fonts.clear();
        number_formats.clear();
        protections.clear();
        
        colors.clear();
    }

	conditional_format add_conditional_format_rule(worksheet_impl *ws, const range_reference &ref, const condition &when)
	{
		conditional_format_impls.push_back(conditional_format_impl());

		auto &impl = conditional_format_impls.back();
		impl.when = when;
		impl.parent = this;
		impl.target_sheet = ws;
		impl.target_range = ref;
		impl.differential_format_id = conditional_format_impls.size() - 1;

		return xlnt::conditional_format(&impl);
	}

    workbook *parent;

    bool operator==(const stylesheet& rhs) const
    {
        // no equality on parent as there is only 1 stylesheet per borkbook hence would always be false
        return garbage_collection_enabled == rhs.garbage_collection_enabled
            && known_fonts_enabled == rhs.known_fonts_enabled
            && conditional_format_impls == rhs.conditional_format_impls
            && format_impls == rhs.format_impls
            && style_impls == rhs.style_impls
            && style_names == rhs.style_names
            && default_slicer_style == rhs.default_slicer_style
            && alignments == rhs.alignments
            && borders == rhs.borders
            && fills == rhs.fills
            && fonts == rhs.fonts
            && number_formats == rhs.number_formats
            && protections == rhs.protections
            && colors == rhs.colors;
    }
    
    bool garbage_collection_enabled = true;
    bool known_fonts_enabled = false;

	std::list<conditional_format_impl> conditional_format_impls;
  
	std::list<format_impl> format_impls;
	// std::unordered_map<std::string, format_impl> format_impls;


    std::unordered_map<std::string, style_impl> style_impls;
    std::vector<std::string> style_names;
    optional<std::string> default_slicer_style;
	optional<std::string> defaultTimelineStyle;
	std::vector<alignment> alignments;
    std::vector<border> borders;
    std::vector<fill> fills;
    std::vector<font> fonts;
    std::vector<number_format> number_formats;
	std::vector<protection> protections;
    
    std::vector<color> colors;
};

} // namespace detail
} // namespace xlnt

**Reply by doomlaur** [reply]
Hello @yang8570! I'm glad you are happy with XLNT. I cannot speak Korean, so I can only translate your post. I'm not sure if your post was meant to be sent as a pull request? If so, please check [CONTRIBUTING.md](https://github.com/xlnt-community/xlnt/blob/master/CONTRIBUTING.md) for a guide on how to contribute to this repository.

If I understand correctly, the code you have posted changes the creation of formats, but I currently do not understand why. You can also explain it in a pull request, if you prefer.

If you choose to send us a pull request, please explain why the change is necessary. Otherwise, if you have any questions, about why formats are working this way, feel free to ask.

**Reply by yang8570** [reply]
나는github를 잘모른다.
단지 원하는것을 찾기 위하여 들렀다

감사합니다.
나는 원하는 조건에 따라서 cell의 색상을 바꿀수 있는 기능이 필요하였다.

엑셀화일을 zip화일로 바꿔서 실제로 xl/styles.xml과 xl\worksheets\sheet2.xml를 검토했습니다.
xlnt::format fmtyellow = ws.cell(c, r).format();
fmtyellow.fill(xlnt::pattern_fill().type(xlnt::pattern_fill_type::solid).foreground(xlnt::color::yellow()));
ws.cell(c, r).format(fmtyellow);
나는 셀의 색상을 채우기만 필요했습니다.
이런작업을 하고나서 기존의 코드에서는 원하지 않는곳의 cell의 색상및 타입도 원하지 않는 방식으로 바뀌었다.
그래서 코드를 검토 하여보니
xl/styles.xml에 정의 되어 있는 <cellXfs count="267">의 항목은 unique하지 않았다.
작업한 순서대로 순차적으로 작업한것을 차곡차곡 쌓아놓았다.
그래서 내나름대로 방법을 찾았다.
나는 내작업을 하기위하여 코드를 수정하였지만 다른 버그는 존재 하는지에 대하여 컴토하지 못하였다.

source\detail\implementations\stylesheet.hpp을 조금 수정 했습니다.
다른곳도 몇군데 손보았지만 내용이 너무 장황해서 이렇게만 이야기를 마치겠다.

**Comment by yang8570** [comment]
-----Original Message-----
